core/context/sourceable.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package context
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"gopkg.in/tomb.v2"
+)
+
+// SourceableError is an interface that can be implemented by errors
+// that can be used as the source of a context error.
+type SourceableError interface {
+	Err() error
+}
+
+type sourceContext struct {
+	ctx    context.Context
+	source SourceableError
+}
+
+// WithSourceableError returns a new Context that carries the source.
+// When the Err method is checked on the context, it will return the error
+// from the source if the context was canceled or deadline exceeded.
+func WithSourceableError(ctx context.Context, source SourceableError) context.Context {
+	return &sourceContext{
+		ctx:    ctx,
+		source: source,
+	}
+}
+
+// Deadline returns the time when work done on behalf of this context
+// should be canceled. Deadline returns ok==false when no deadline is
+// set. Successive calls to Deadline return the same results.
+func (t *sourceContext) Deadline() (deadline time.Time, ok bool) {
+	return t.ctx.Deadline()
+}
+
+// Done returns a channel that's closed when work done on behalf of this
+// context should be canceled. Done may return nil if this context can
+// never be canceled. Successive calls to Done return the same value.
+// The close of the Done channel may happen asynchronously,
+// after the cancel function returns.
+func (t *sourceContext) Done() <-chan struct{} {
+	return t.ctx.Done()
+}
+
+// If Done is not yet closed, Err returns nil.
```

Comment:
Comment for doesn't start with `Err`.

Reply:
I lifted this from the go docs on context :wink: